pkg/util/unpacker.go

```diff
@@ -0,0 +1,101 @@
+package source
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	rukpakv1alpha1 "github.com/operator-framework/rukpak/api/v1alpha1"
+	bundlesource "github.com/operator-framework/rukpak/internal/source"
+	pkgsource "github.com/operator-framework/rukpak/pkg/source"
+)
+
+const (
+	// uploadClientTimeout is the timeout to be used with http connections to upload manager.
+	uploadClientTimeout = time.Second * 10
+)
+
+type unpacker struct {
+	sources map[rukpakv1alpha1.SourceType]pkgsource.Unpacker
+}
+
+// NewUnpacker returns a new composite Source that unpacks bundles using the source
+// mapping provided by the configured sources.
+func NewUnpacker(sources map[rukpakv1alpha1.SourceType]pkgsource.Unpacker) pkgsource.Unpacker {
+	return &unpacker{sources: sources}
+}
+
+func (s *unpacker) Unpack(ctx context.Context, bundle *rukpakv1alpha1.Bundle) (*pkgsource.Result, error) {
+	source, ok := s.sources[bundle.Spec.Source.Type]
+	if !ok {
+		return nil, fmt.Errorf("source type %q not supported", bundle.Spec.Source.Type)
+	}
+	return source.Unpack(ctx, bundle)
+}
+
+type Config struct {
+	// ConfigMgr is key of the config for ctrl.Manager
+	Manager ctrl.Manager
+	// ConfigNamespace is key of the config for namespace string
+	Namespace string
+	// ConfigUnpackImage is key of the config for unpack image name string
+	UnpackImage string
+	// ConfigBaseUploadManagerURL is key of the config for base upload manager URL string
+	BaseUploadManagerURL string
+	// ConfigRootCAs is key of the config for *x509.CertPool value for Root CAs
+	RootCAs *x509.CertPool
+}
```

Comment:
This struct-based approach was discussed, as was the functional options pattern in #478.

The reason we use individual function parameters now is that every field is required. And the reason we're concerned with that function signature is because that list of parameters is growing. We've been adding more required fields.

With the current approach on `main`, we break the API by adding a new parameter, thus causing the compiler to complain at compile time.

With this approach, we'd add a new field to the struct, and existing code would still compile even if it was using the new struct definition. And a provisioner developer wouldn't find out until runtime that something went wrong. Maybe not the end of the world, but less good than what we have now.

FWIW, this Go language proposal for required struct fields resulted in the recommendation of "just use a constructor", which is what we're doing now.

So perhaps our existing `NewDefaultUnpacker` function is actually fine?

Reply:
OK.  I'll put the `NewDefaultUnpacker` back.  So the provisioners must be updated and rebuilt when the new source with additional configuration added.